File: partner-frontend/src/shared/types/order.ts
/**
 * Canonical QuickPress order contract.
 *
 * This is the ONE order model every application (customer, partner, rider,
 * admin) agrees on. Each app has its own view type tuned to its screens; those
 * are derived from this model by the mappers in `@backend/mock/mappers`, never
 * duplicated.
 *
 * When the real FastAPI backend lands, these are the response bodies the API
 * must return — nothing in the UI changes.
 */

/** The complete order lifecycle, in progression order. */
export type OrderLifecycleStatus =
  | "placed" // customer placed the order, partner has it in "new"
  | "partner_accepted" // partner accepted the order
  | "rider_assigned" // a rider was assigned for pickup
  | "picked_up" // rider collected laundry from the customer
  | "at_partner" // laundry handed over at the partner facility
  | "processing" // partner is washing / cleaning
  | "completed" // partner marked laundry completed (ready for delivery)
  | "out_for_delivery" // rider is delivering back to the customer
  | "delivered" // customer received the order
  | "cancelled";

export const ORDER_LIFECYCLE: OrderLifecycleStatus[] = [
  "placed",
  "partner_accepted",
  "rider_assigned",
  "picked_up",
  "at_partner",
  "processing",
  "completed",
  "out_for_delivery",
  "delivered",
];

export const ORDER_STATUS_LABEL: Record<OrderLifecycleStatus, string> = {
  placed: "Order placed",
  partner_accepted: "Accepted by store",
  rider_assigned: "Rider assigned",
  picked_up: "Picked up",
  at_partner: "Reached store",
  processing: "In cleaning",
  completed: "Laundry completed",
  out_for_delivery: "Out for delivery",
  delivered: "Delivered",
  cancelled: "Cancelled",
};

export function orderStageIndex(status: OrderLifecycleStatus): number {
  const index = ORDER_LIFECYCLE.indexOf(status);
  return index < 0 ? 0 : index;
}

export function isOrderAfter(
  status: OrderLifecycleStatus,
  reference: OrderLifecycleStatus,
): boolean {
  if (status === "cancelled") return false;
  return orderStageIndex(status) >= orderStageIndex(reference);
}

export type OrderParty = {
  id: string;
  name: string;
  phone: string;
};

export type OrderRiderParty = OrderParty & {
  vehicle: string;
  plate: string;
  rating: number;
  trips: string;
};

export type OrderPartnerParty = OrderParty & {
  image: string;
  city: string;
};

export type OrderLine = {
  id: string;
  name: string;
  qty: number;
  price: number;
};

export type OrderAddress = {
  label: string;
  line: string;
  city: string;
  phone: string;
};

export type OrderTotals = {
  itemsTotal: number;
  pickup: number;
  delivery: number;
  handling: number;
  gst: number;
  discount: number;
  grandTotal: number;
};

export type OrderEvent = {
  id: string;
  status: OrderLifecycleStatus;
  label: string;
  /** ISO timestamp. */
  at: string;
  /** Which app/actor moved the order forward. */
  actor: "customer" | "partner" | "rider" | "admin" | "system";
};

/** The single shared order entity. */
export type Order = {
  id: string;
  code: string;
  status: OrderLifecycleStatus;
  createdAt: string;
  updatedAt: string;
  customer: OrderParty;
  partner: OrderPartnerParty;
  rider: OrderRiderParty | null;
  serviceLabel: string;
  items: OrderLine[];
  totals: OrderTotals;
  address: OrderAddress;
  pickup: { date: string; slot: string; express: boolean };
  delivery: { date: string; slot: string };
  payment: { mode: "online" | "cod"; label: string; note: string; paid: boolean };
  /** OTPs the rider collects at pickup and at delivery. */
  otp: { pickup: string; delivery: string };
  events: OrderEvent[];
  cancelledReason: string | null;
};

export type PlaceOrderPayload = {
  /** Optional: the backend resolves it from the bearer token when omitted. */
  customerId?: string;
  partnerId?: string;
  items: OrderLine[];
  addressId?: string;
  address?: OrderAddress;
  pickup: { date: string; slot: string; express: boolean };
  delivery?: { date: string; slot: string };
  payment: { mode: "online" | "cod"; label: string; note?: string };
  totals?: Partial<OrderTotals>;
  serviceLabel?: string;
};

File: backend-python/app/models/order.py
"""Order models — Sprint 2.4 (Checkout & Order Creation).

These mirror the canonical `Order` contract in `shared/src/types/order.ts`, so
POST /api/orders and GET /api/orders/{order_id} feed the existing Order Success
and Live Tracking screens without any UI change.
"""

from __future__ import annotations

from typing import List, Literal, Optional

from pydantic import BaseModel

from app.models.address import AddressType  # noqa: F401  (kept for callers)

OrderStatus = Literal[
    # Canonical lifecycle statuses (shared by customer, partner, rider, admin).
    "pending_partner_acceptance",
    "placed",  # legacy alias of pending_partner_acceptance, kept for old rows
    "partner_accepted",
    "rider_accepted",
    "rider_assigned",
    "picked_up",
    "at_partner",
    "processing",
    "completed",
    "out_for_delivery",
    "delivered",
    "cancelled",
]


class OrderLine(BaseModel):
    id: str
    name: str
    qty: int
    price: int


class OrderAddress(BaseModel):
    label: str = "Home"
    line: str = ""
    city: str = ""
    phone: str = ""


class OrderTotals(BaseModel):
    itemsTotal: int = 0
    pickup: int = 0
    delivery: int = 0
    handling: int = 0
    gst: int = 0
    discount: int = 0
    grandTotal: int = 0


class OrderPickup(BaseModel):
    date: str = ""
    slot: str = ""
    express: bool = False


class OrderDelivery(BaseModel):
    date: str = ""
    slot: str = ""


class OrderPaymentPayload(BaseModel):
    mode: Literal["online", "cod"] = "cod"
    label: str = "Cash on delivery"
    note: Optional[str] = None
    method: Optional[str] = None


class OrderPayment(BaseModel):
    mode: Literal["online", "cod"] = "cod"
    label: str = "Cash on delivery"
    note: str = ""
    paid: bool = False


class OrderParty(BaseModel):
    id: str = ""
    name: str = ""
    phone: str = ""


class OrderPartnerParty(OrderParty):
    image: str = ""
    city: str = ""


class OrderRiderParty(OrderParty):
    vehicle: str = ""
    plate: str = ""
    rating: float = 0
    trips: str = ""


class OrderEvent(BaseModel):
    id: str
    status: OrderStatus
    label: str
    at: str
    actor: Literal["customer", "partner", "rider", "admin", "system"] = "customer"


class OrderOtp(BaseModel):
    pickup: str = ""
    delivery: str = ""


class PlaceOrderPayload(BaseModel):
    """POST /api/orders."""

    items: List[OrderLine] = []
    addressId: Optional[str] = None
    address: Optional[OrderAddress] = None
    partnerId: Optional[str] = None
    serviceLabel: Optional[str] = None
    pickup: OrderPickup = OrderPickup()
    delivery: Optional[OrderDelivery] = None
    payment: OrderPaymentPayload = OrderPaymentPayload()
    totals: Optional[OrderTotals] = None
    couponCode: Optional[str] = None
    couponDiscount: int = 0
    instructions: str = ""
    """Client generated key so a double tap can never create two orders."""
    idempotencyKey: Optional[str] = None


class OrderResponse(BaseModel):
    """The canonical order document — identical shape in mock and FastAPI."""

    id: str
    code: str
    status: OrderStatus = "pending_partner_acceptance"
    createdAt: str
    updatedAt: str
    customer: OrderParty = OrderParty()
    partner: OrderPartnerParty = OrderPartnerParty()
    rider: Optional[OrderRiderParty] = None
    serviceLabel: str = "Laundry"
    items: List[OrderLine] = []
    totals: OrderTotals = OrderTotals()
    address: OrderAddress = OrderAddress()
    pickup: OrderPickup = OrderPickup()
    delivery: OrderDelivery = OrderDelivery()
    payment: OrderPayment = OrderPayment()
    otp: OrderOtp = OrderOtp()
    events: List[OrderEvent] = []
    cancelledReason: Optional[str] = None


class PlaceOrderResponse(BaseModel):
    """POST /api/orders — what the checkout screen needs to move on."""

    orderId: str
    orderNumber: str
    pickupEstimate: str
    deliveryEstimate: str
    order: OrderResponse
